cupy/core/raw.pyx

```diff
@@ -72,41 +74,29 @@ def _get_raw_kernel(code, name, options=()):
     return module.get_function(name)
 
 
-@cupy.util.memoize(for_each_device=True)
-def _get_func_attributes(func):
-    cdef:
-        int sharedSizeBytes, constSizeBytes, localSizeBytes
-        int maxThreadsPerBlock, numRegs, ptxVersion, binaryVersion
-        int cacheModeCA, maxDynamicSharedSizeBytes, preferredShmemCarveout
-
-    sharedSizeBytes = driver.funcGetAttribute(
-        driver.CU_FUNC_ATTRIBUTE_SHARED_SIZE_BYTES, func)
-    constSizeBytes = driver.funcGetAttribute(
-        driver.CU_FUNC_ATTRIBUTE_CONST_SIZE_BYTES, func)
-    localSizeBytes = driver.funcGetAttribute(
-        driver.CU_FUNC_ATTRIBUTE_LOCAL_SIZE_BYTES, func)
-    maxThreadsPerBlock = driver.funcGetAttribute(
-        driver.CU_FUNC_ATTRIBUTE_MAX_THREADS_PER_BLOCK, func)
-    numRegs = driver.funcGetAttribute(
-        driver.CU_FUNC_ATTRIBUTE_NUM_REGS, func)
-    ptxVersion = driver.funcGetAttribute(
-        driver.CU_FUNC_ATTRIBUTE_PTX_VERSION, func)
-    binaryVersion = driver.funcGetAttribute(
-        driver.CU_FUNC_ATTRIBUTE_BINARY_VERSION, func)
-    cacheModeCA = driver.funcGetAttribute(
-        driver.CU_FUNC_ATTRIBUTE_CACHE_MODE_CA, func)
-    maxDynamicSharedSizeBytes = driver.funcGetAttribute(
-        driver.CU_FUNC_ATTRIBUTE_MAX_DYNAMIC_SHARED_SIZE_BYTES, func)
-    preferredShmemCarveout = driver.funcGetAttribute(
-        driver.CU_FUNC_ATTRIBUTE_PREFERRED_SHARED_MEMORY_CARVEOUT, func)
-
-    return dict(sharedSizeBytes=sharedSizeBytes,
-                constSizeBytes=constSizeBytes,
-                localSizeBytes=localSizeBytes,
-                maxThreadsPerBlock=maxThreadsPerBlock,
-                numRegs=numRegs,
-                ptxVersion=ptxVersion,
-                binaryVersion=binaryVersion,
-                cacheModeCA=cacheModeCA,
-                maxDynamicSharedSizeBytes=maxDynamicSharedSizeBytes,
-                preferredShmemCarveout=preferredShmemCarveout)
+class FunctionAttributes:
+
+    class Read:
+        def __init__(self, func_attribute):
+            self.func_attribute = func_attribute
+
+        def __get__(self, instance, owner):
+            return driver.funcGetAttribute(self.func_attribute, instance.kern.ptr)
+
+    class ReadWrite(Read):
+        def __set__(self, instance, value):
+            driver.funcSetAttribute(instance.kern.ptr, self.func_attribute, value)
+
+    shared_size_bytes = Read(driver.CU_FUNC_ATTRIBUTE_SHARED_SIZE_BYTES)
```

Comment:
Although this underscore format is more Pythonic, I think the core devs had previously expressed preference for using variable names that closely match those from the CUDA libs. I would leave it as is for now pending feedback, though.

Reply:
To add to the confusion, the runtime and driver APIs use different conventions .. as seen here, the driver API uses all-caps and underscore delimiters. So I just changed the all caps to lowercase which, as you noted, yields standard python style names.

Reply:
I have a different question related to `__set__`: do we want to do a foolproof here? For example, if I understand it correctly due to hardware limitation we have `maxDynamicSharedSizeBytes` <= `sharedMemPerMultiprocessor`, so perhaps when setting the former we should query the device property, available from #1869, and check if we ask too much?

Reply:
The CUDA driver API returns an error if you request too much shared memory, and `funcSetAttribute` handles all errors via `check_status`. 

In this case, the the setter raises a `cupy.cuda.driver.CUDADriverError` with value `CUDA_ERROR_INVALID_VALUE`.

Reply:
That's right. This is indeed what I observed in https://github.com/cupy/cupy/issues/2118#issuecomment-479546034. My concern is that the raised error is not very verbose, so the user may not be able to figure out quickly the cause.

Reply:
Just to be clear I am saying if you set the `max_dynamic_shared_size_bytes` property to an invalid value, then a `CUDA_ERROR_INVALID_VALUE` exception is raised. Since ultimately the driver decides which values are valid, I would be wary about validating the value ourselves. Maybe the question is whether we should rewrite the error message to make it more specific to the attribute the user is setting.

Reply:
I agree that `CUDA_ERROR_INVALID_VALUE` probably gives a good enough clue about values being out of range. 

However should we be at least checking that the user provided `value` is `int`-like? (As a concrete example, I can see a user wanting a 30% shared memory carveout and not realize the value specified should be an integer in range [0, 100] and instead providing a fraction such as 0.3)?

Reply:
I see .. my philosophy is usually to err on the side of simplicity and write good documentation (though I might need some encouragement for the latter).

Reply:
Added some attribute documentation.